packages/next/server/image-optimizer.ts

```diff
@@ -473,15 +477,35 @@ async function writeToCacheDir(
   await promises.writeFile(filename, buffer)
 }
 
+function getFileNameWithExtension(
+  url: string,
+  contentType: string | null
+): string | void {
+  const [urlWithoutQueryParams] = url.split('?')
+  const fileNameWithExtension = urlWithoutQueryParams.split('/').pop()
+  if (!contentType || !fileNameWithExtension) {
+    return
+  }
+
+  const [fileName] = fileNameWithExtension.split('.')
+  const extension = getExtension(contentType)
+  return `${fileName}.${extension}`
+}
+
 function setResponseHeaders(
   req: IncomingMessage,
   res: ServerResponse,
+  url: string,
   etag: string,
   maxAge: number,
   contentType: string | null,
   isStatic: boolean,
   isDev: boolean
 ) {
+  const fileName = getFileNameWithExtension(url, contentType)
+  if (fileName) {
+    res.setHeader('Content-Disposition', `inline; filename="${fileName}"`)
+  }
```

Comment:
I don't think we need to set the header in the case of a 304

Lets move this below `Content-Type`